etcdserver/v3demo_server.go

```diff
@@ -127,29 +138,41 @@ func applyTxn(kv dstorage.KV, rt *pb.TxnRequest) *pb.TxnResponse {
 	txnResp.Header.Revision = revision
 	txnResp.Responses = resps
 	txnResp.Succeeded = ok
-	return txnResp
+	return txnResp, nil
 }
 
-func applyCompaction(kv dstorage.KV, compaction *pb.CompactionRequest) *pb.CompactionResponse {
+func applyCompaction(kv dstorage.KV, compaction *pb.CompactionRequest) (*pb.CompactionResponse, error) {
 	resp := &pb.CompactionResponse{}
 	resp.Header = &pb.ResponseHeader{}
 	err := kv.Compact(compaction.Revision)
 	if err != nil {
-		panic("handle error")
+		return nil, err
 	}
 	// get the current revision. which key to get is not important.
 	_, resp.Header.Revision, _ = kv.Range([]byte("compaction"), nil, 1, 0)
-	return resp
+	return resp, err
 }
 
 func applyUnion(kv dstorage.KV, union *pb.RequestUnion) *pb.ResponseUnion {
 	switch {
 	case union.RequestRange != nil:
-		return &pb.ResponseUnion{ResponseRange: applyRange(kv, union.RequestRange)}
+		resp, err := applyRange(kv, union.RequestRange)
+		if err != nil {
+			panic("unexpected error during txn")
```

Comment:
we should check potential errors before executing the transaction. we cannot abort executed commands. 

see line 120 for more details.